app/lib/activitypub/activity/announce.rb

```diff
@@ -34,6 +29,84 @@ def perform
 
   private
 
+  def process_status
+    @mentions = []
+    @params   = {}
+
+    process_status_params
+    process_audience
+
+    ApplicationRecord.transaction do
+      @status = Status.create!(@params)
+      attach_mentions(@status)
+    end
+
+    distribute(@status)
+  end
+
+  def process_status_params
+    @params = begin
+      {
+        account: @account,
+        reblog: @original_status,
+        uri: @json['id'],
+        created_at: @json['published'],
+        override_timestamps: @options[:override_timestamps],
+        visibility: visibility_from_audience
+      }
+    end
+  end
+
+  def process_audience
+    conversation_uri = value_or_id(@object['context'])
+
+    (audience_to + audience_cc).uniq.each do |audience|
+      next if audience == ActivityPub::TagManager::COLLECTIONS[:public] || audience == conversation_uri
+
+      # Unlike with tags, there is no point in resolving accounts we don't already
+      # know here, because silent mentions would only be used for local access
+      # control anyway
+      account = account_from_uri(audience)
+
+      next if account.nil? || @mentions.any? { |mention| mention.account_id == account.id }
+
+      @mentions << Mention.new(account: account, silent: true)
+
+      # If there is at least one silent mention, then the status can be considered
+      # as a limited-audience status, and not strictly a direct message, but only
+      # if we considered a direct message in the first place
+      @params[:visibility] = :limited if @params[:visibility] == :direct
+    end
+
+    # If the payload was delivered to a specific inbox, the inbox owner must have
+    # access to it, unless they already have access to it anyway
+    return if @options[:delivered_to_account_id].nil? || @mentions.any? { |mention| mention.account_id == @options[:delivered_to_account_id] }
+
+    @mentions << Mention.new(account_id: @options[:delivered_to_account_id], silent: true)
+
+    @params[:visibility] = :limited if @params[:visibility] == :direct
+  end
+
+  def postprocess_audience_and_deliver
+    return if @status.mentions.find_by(account_id: @options[:delivered_to_account_id])
+
+    delivered_to_account = Account.find(@options[:delivered_to_account_id])
+
+    @status.mentions.create(account: delivered_to_account, silent: true)
+    @status.update(visibility: :limited) if @status.direct_visibility?
```

Comment:
Not a big deal, but not useful here as reblogs can't have `direct` visibility.

Reply:
Moved duplicate code for Create and Announce to one place.